dd-sdk-android/src/main/kotlin/com/datadog/android/sessionreplay/internal/domain/SessionReplayRequestFactory.kt

```diff
@@ -6,41 +6,178 @@
+    private val requestBodyFactory: RequestBodyFactory = RequestBodyFactory(),
+    internal val internalLogger: Logger = sdkLogger,
+    private val bodyToByteArrayFactory: (RequestBody) -> ByteArray = {
+        val buffer = Buffer()
+        // this is handled below in the wrapper function
+        @Suppress("UnsafeThirdPartyFunctionCall")
+        it.writeTo(buffer)
+        @Suppress("UnsafeThirdPartyFunctionCall")
+        buffer.readByteArray()
+    }
 ) : RequestFactory {
     override fun create(
         context: DatadogContext,
         batchData: List<ByteArray>,
         batchMetadata: ByteArray?
     ): Request {
-        // TODO: RUMM-2547 Drop this code and return a list of requests instead when
-        // the feature will be available in SDK V2
-        // Also add the necessary unit tests once this is done.
-        batchToSegmentsMapper.map(batchData).forEach {
-            sessionReplayOkHttpUploader.upload(
-                context,
-                it.first,
-                (it.second.toString() + "\n").toByteArray()
-            )
+        val serializedSegmentPair = batchToSegmentsMapper.map(batchData)
+        if (serializedSegmentPair == null) {
+            // TODO: RUMM-2397 Add the proper logs here once the sdkLogger will be added
+            // the data in the batch is corrupted. We return an empty request just to make
+            // sure the batch will be deleted by the core.
+            return getEmptyRequest()
         }
+        val body = requestBodyFactory.create(
+            serializedSegmentPair.first,
+            serializedSegmentPair.second
+        )
+        return resolveRequest(context, body)
+    }
+
+    private fun getEmptyRequest(): Request {
+        val id = UUID.randomUUID().toString()
         return Request(
-            UUID.randomUUID().toString(),
+            id,
             "",
-            sessionReplayOkHttpUploader.buildUrl(context),
-            mapOf(
-                DataOkHttpUploaderV2.HEADER_API_KEY to context.clientToken
-            ),
-            ByteArray(0)
+            "",
+            emptyMap(),
+            body = ByteArray(0)
         )
     }
+
+    private val intakeUrl by lazy {
+        String.format(
+            Locale.US,
+            UPLOAD_URL,
+            endpoint,
+            "replay"
+        )
+    }
+
+    private fun tags(datadogContext: DatadogContext): String {
+        val elements = mutableListOf(
+            "${RumAttributes.SERVICE_NAME}:${datadogContext.service}",
+            "${RumAttributes.APPLICATION_VERSION}:" +
+                datadogContext.version,
+            "${RumAttributes.SDK_VERSION}:${datadogContext.sdkVersion}",
+            "${RumAttributes.ENV}:${datadogContext.env}"
+        )
+        if (datadogContext.variant.isNotEmpty()) {
+            elements.add("${RumAttributes.VARIANT}:${datadogContext.variant}")
+        }
+        return elements.joinToString(",")
```

Comment:
are these tags are still applicable for the SR? `RumAttributes` will be in the RUM module eventually, but we can make them a part of public API.